rustsat/src/instances/fio.rs

```diff
@@ -66,3 +75,82 @@ pub(crate) fn open_compressed_uncompressed_write<P: AsRef<Path>>(
     }
     Ok(Box::new(io::BufWriter::new(raw_writer)))
 }
+
+pub enum SolverOutput {
+    Sat(types::Assignment),
+    Unsat,
+    Unknown,
+}
+
+#[derive(Error, Debug)]
+pub enum SatSolverOutputError {
+    //The output of the SAT solver is incorrect.
+    #[error("The output of the SAT solver is incorrect.")]
+    Nonsolution,
+    #[error("No solution line found in the output.")]
+    NoSline,
+    #[error("No value line found in the output.")]
+    NoVline,
+    #[error("Invalid solution line found in the output.")]
+    InvalidSLine,
+}
+
+//Parse a SAT solver's output
+pub fn parse_sat_solver_output<R: BufRead>(reader: R) -> anyhow::Result<SolverOutput> {
+    // Parsing code goes here
+    // when encoutering vline, call `Assignment::from_vline`
+
+    let mut is_sat = false;
+    let mut have_vline = false;
+    let mut solution = None;
+
+    for line in reader.lines() {
+        let line = &line?;
+        //let line = &line;
+
+        //Solution line
+        if line.starts_with('s') {
+            match line {
+                line if line.contains("UNSATISFIABLE") => return Ok(SolverOutput::Unsat),
+                line if line.contains("UNKNOWN") => {
+                    return Err(anyhow::anyhow!(SatSolverOutputError::InvalidSLine))
+                }
+                line if line.contains("SATISFIABLE") => {
+                    if let Some(solution) = solution {
+                        return Ok(SolverOutput::Sat(solution));
+                    }
+                    is_sat = true;
+                }
+                _ => return Err(anyhow::anyhow!(SatSolverOutputError::InvalidSLine)),
+            }
+        }
+
+        //Value line
+        if line.starts_with('v') {
+            if have_vline {
+                let mut current_assignment = match solution {
+                    Some(ref mut assign) => assign,
+                    _ => return Err(anyhow::anyhow!(SatSolverOutputError::Nonsolution)),
+                };
+
+                current_assignment.extend_from_vline(&line)?;
+            } else {
+                solution = Some(Assignment::from_vline(&line)?);
+            }
+
+            have_vline = true;
+        }
+    }
+
+    if have_vline {
+        if let Some(solution) = solution {
+            return Ok(SolverOutput::Sat(solution));
+        }
+    }
+
+    if is_sat {
+        Err(anyhow::anyhow!(SatSolverOutputError::NoSline))
+    } else {
+        Err(anyhow::anyhow!(SatSolverOutputError::NoVline))
+    }
+}
```

Comment:
```suggestion
}


#[cfg(test)]
mod tests {
    use std::io;

    use crate::types::{Assignment, TernaryVal};

    use super::{parse_sat_solver_output, SolverOutput};

    #[test]
    fn parse_solver_output_sat() {
        let ground_truth = SolverOutput::Sat(Assignment::from(vec![
            TernaryVal::True,
            TernaryVal::False,
            TernaryVal::DontCare,
            TernaryVal::True,
            TernaryVal::False,
            TernaryVal::True,
        ]));

        let data = "c this is a comment\ns SATISFIABLE\nv 1 -2 4 -5 6 0\n";
        let reader = io::Cursor::new(data);
        let res = parse_sat_solver_output(reader).unwrap();
        assert_eq!(res, ground_truth);

        let data = "c this is a comment\nv 1 -2 4 -5 6 0\ns SATISFIABLE\n";
        let reader = io::Cursor::new(data);
        let res = parse_sat_solver_output(reader).unwrap();
        assert_eq!(res, ground_truth);

        let data = "c this is a comment\ns SATISFIABLE\nv 1 -2 4 \nv -5 6 0\n";
        let reader = io::Cursor::new(data);
        let res = parse_sat_solver_output(reader).unwrap();
        assert_eq!(res, ground_truth);
    }

    #[test]
    fn parse_solver_output_unsat() {
        let data = "c this is a comment\ns UNSATISFIABLE\n";
        let reader = io::Cursor::new(data);
        let res = parse_sat_solver_output(reader).unwrap();
        assert_eq!(res, SolverOutput::Unsat);
    }

    #[test]
    fn parse_solver_output_unknown() {
        let data = "c this is a comment\ns UNKNOWN\n";
        let reader = io::Cursor::new(data);
        let res = parse_sat_solver_output(reader).unwrap();
        assert_eq!(res, SolverOutput::Unknown);
    }
}
```

Reply:
Just some more suggested tests I've messed up myself over the years:

A completely empty SAT file (which people sometimes generate automatically, so should be handled cleanly): `s SATISFIABLE\nv 0\n`